src/mame/tandberg/tdv2100_disp_logic.cpp

```diff
@@ -0,0 +1,1029 @@
+}
+
+void tandberg_tdv2100_disp_logic_device::device_reset()
+{
+	m_cursor_row_input = false;
+	m_cursor_col_input = false;
+	m_page_roll = 0;
+	char_to_display(0x19);
+	m_frame_counter = 0;
+	m_underline_input = false;
+	m_video_enable = true;
+	m_data_terminal_ready = false;
+	m_request_to_send = false;
+	m_data_pending_kbd = false;
+	m_uart->write_xr(1);
+	m_uart->write_xr(0);
+	w_cleark(1);
+	w_cleark(0);
+
+	bool force_on_line = m_dsw_u39->read()&0x020;
+	set_uart_state_from_switches();
+	if(force_on_line)
+	{
+		w_linek(1);
+		w_linek(0);
+		w_transk(1);
+		w_transk(0);
+	}
+	update_rs232_lamps();
+}
+
+///////////////////////////////////////////////////////////////////////////////
+//
+// Graphics and character data
+//
+
+void tandberg_tdv2100_disp_logic_device::char_to_display(uint8_t byte)
+{
+	// Note: There are no checks in hardware to catch an invalid cursor-position.
+	//       To clear an invalid cursor placement, the cursor-movement mechanism
+	//       may eventually catch it later, but it can also be manually sorted
+	//       out by pressing the CR and/or HOME key.
+	if(m_cursor_row_input)
+	{
+		place_cursor(byte, m_cursor_col);
+		m_cursor_col_input = true;
+		m_cursor_row_input = false;
+	}
+	else if(m_cursor_col_input)
+	{
+		place_cursor(m_cursor_row, byte);
+		m_cursor_col_input = false;
+	}
+	else
+	{
+		// Ordinary input
+		if((byte&0x7f) < 0x20 || (byte&0x7f) == 0x7f)
+		{
+			// Control characters
+			bool auto_roll_up = !(m_dsw_u39->read()&0x100);
+			bool clear_ascii_hs_by_cleark = ((m_dsw_u39->read()&0x003) == 0x001);
+			switch(byte&0x7f)
+			{
+				// STX, Turn off video
+				case 0x02:
+					m_video_enable = false;
+					break;
+
+				// ETX, Turn on video
+				case 0x03:
+					m_video_enable = true;
+					break;
+
+				// UL, Turn on Underline on new input
+				case 0x0e:
+					m_underline_input = true;
+					break;
+
+				// NO, Normal input (Underline on input off)
+				case 0x0f:
+					m_underline_input = false;
+					break;
+
+				// ERLIN, Erase line and move cursor to line start
+				case 0x04:
+					erase_row(m_cursor_row);
+					char_to_display(0x0d);
+					break;
+
+				// ERPAG, Clear screen and move cursor home
+				case 0x19:
+					for(int row=0; row<25; row++)
+					{
+						erase_row(row);
+					}
+					char_to_display(0x1d);
+					break;
+
+				// CURL, Cursor left
+				case 0x08:
+					if(m_cursor_col > 0)
+					{
+						place_cursor(m_cursor_row, m_cursor_col-1);
+					}
+					break;
+
+				// CURR, Cursor right
+				case 0x18:
+					advance_cursor();
+					break;
+
+				// CURUP, Cursor up
+				case 0x1c:
+					if(m_cursor_row > 0)
+					{
+						place_cursor(m_cursor_row-1, m_cursor_col);
+					}
+					break;
+
+				// CURD, Cursor down
+				case 0x0b:
+					if((m_cursor_row&0x18) != 24)
+					{
+						place_cursor(m_cursor_row+1, m_cursor_col);
+					}
+					break;
+
+				// LF, Move cursor to next line
+				case 0x0a:
+					if((m_cursor_row&0x18) == 24 && auto_roll_up)
```

Comment:
It would be clearer if you keep it in hex when doing comparisons, and most of the code has more white space which seems more readable to me. `if (((m_cursor_row & 0x18) == 0x18) && auto_roll_up)`

Reply:
Given what this logic is responsible of, I thought it would be more readable to have the it this way. 0x18 mask which lines are used for the comparison, 24 is the row-number the comparison targets. Of course the target uses all the lines of the mask by nature, but it can be useful to be able to directly read out which row is targeted without having to convert from hex.